src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh

```diff
@@ -0,0 +1,410 @@
+
+    real_type skin_depth_;
+    MevMass neutron_mass_;
+    MevMass proton_mass_;
+
+    CollectionBuilder<ZoneComponent> components_;
+    CollectionBuilder<NuclearZones, MemSpace::host, IsotopeId> zones_;
+
+    //// COMMON PROPERTIES ////
+
+    static CELER_CONSTEXPR_FUNCTION real_type half() { return 0.5; }
+    static CELER_CONSTEXPR_FUNCTION real_type four_thirds_pi()
+    {
+        return 4 * constants::pi / real_type{3};
+    }
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with cascade options and data.
+ */
+CELER_FUNCTION
+NuclearZoneBuilder::NuclearZoneBuilder(CascadeOptions const& options,
+                                       NeutronInelasticScalars const& scalars,
+                                       Data* data)
+    : options_(options)
+    , skin_depth_(0.611207 * options.radius_scale)
+    , neutron_mass_(scalars.neutron_mass)
+    , proton_mass_(scalars.proton_mass)
+    , components_(&data->components)
+    , zones_(&data->zones)
+{
+    CELER_EXPECT(options_);
+    CELER_EXPECT(data);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct nuclear zone data for a single isotope.
+ */
+void NuclearZoneBuilder::operator()(IsotopeView const& target)
+{
+    AtomicNumber a = target.atomic_mass_number();
+    size_type num_zones = this->num_nuclear_zones(a);
+
+    std::vector<ZoneComponent> comp(num_zones);
+    this->calc_zone_component(target, comp);
+
+    NuclearZones nucl_zones;
+    nucl_zones.num_zones = num_zones;
+    nucl_zones.zones = components_.insert_back(comp.begin(), comp.end());
+    zones_.push_back(nucl_zones);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return the number of nuclear zones.
+ */
+size_type NuclearZoneBuilder::num_nuclear_zones(AtomicMassNumber a)
+{
+    return (a.get() < 5) ? 1 : (a.get() < 100) ? 3 : 6;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate components of nuclear zone data: the nuclear zone radius, volume,
+ * density, Fermi momentum and potential function as in G4NucleiModel and as
+ * documented in section 24.2.3 of the Geant4 Physics Reference (release 11.2).
+ */
+CELER_FUNCTION
+void NuclearZoneBuilder::calc_zone_component(IsotopeView const& target,
+                                             ComponentVec& components)
+{
+    AtomicNumber a = target.atomic_mass_number();
+
+    // Calculate nuclear radius
+    real_type nuclear_radius = this->calc_nuclear_radius(a);
+
+    real_type skin_ratio = nuclear_radius / skin_depth_;
+    real_type skin_decay = std::exp(-skin_ratio);
+
+    // Temporary data for the zone-by-zone density function
+    size_type num_of_zones = components.size();
+    std::vector<real_type> ur(num_of_zones + 1);
```

Comment:
I don't think you need a whole vector of `ur`: only `[i, i+1]` are used at a time...